Mk 243862 reassign cases manage forms

mkangia opened the pull request:
Branching out from https://github.com/dimagi/commcare-hq/pull/14206 to make location safe:
1. Reassign Cases
2. Manage Forms

mkangia committed:
Make case reassignment Location Safe

mkangia committed:
Restrict Case reassignment to only accessible users

mkangia committed:
make 'Manage Forms' location safe

esoergel commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -85,12 +87,22 @@ def rows(self):
             users=[dict(ownerid=user.user_id, name=user.username_in_report, type="user")
                    for user in active_users],
-            groups=[dict(ownerid=group.get_id, name=group.name, type="group")
-                    for group in self.all_case_sharing_groups],
```
lol, this old dict format is a throwback, huh?

mkangia replied:
@esoergel whats that?

esoergel replied:
Oh, nothing, just commenting that this uses a `dict` constructor instead of a `dict` literal:
```python
dict(ownerid=group.get_id, name=group.name, type="group")
# vs
{'ownerid': group.get_id, 'name': group.name, 'type': "group"}
```
(Totally fine to leave as-is, it's just unusual code I don't see often)

mkangia replied:
Aha!

esoergel commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -85,12 +87,22 @@ def rows(self):
+            )
+            accessible_user_ids = user_ids_at_locations(accessible_location_ids)
+            active_users = self.get_all_users_by_domain(user_filter=tuple(HQUserType.all()), simplified=True,
+                                                        user_ids=tuple(accessible_user_ids))
```
Non-blocking, but I'm concerned about the performance of this call.  I just looked at the definition of `get_all_users_by_domain`, and it does `CommCareUser.get` for every id.  That means potentially hundreds or thousands of database hits.  It would be faster to do this in bulk.  Something like this should do it:

```python
user_query = UserES().domain(domain).location(accessible_location_ids)
active_users = get_simplified_users(user_query)
```

Or some other way of calling `_report_user_dict`.  (I just ran in to something similar recently and added `get_simplified_users` to make it easy to go from ES queries to report users)

mkangia replied:
Thats right. Thank You @esoergel.
Using the first suggestion. I think `get_all_users_by_domain` just handles the case where a user is not found and gives a temp user but i don't think that would be an issue here.

mkangia replied:
Moreover, is there a way to improve `get_all_users_by_domain` as well?

esoergel replied:
> Moreover, is there a way to improve get_all_users_by_domain as well?
Haha, probably!  There may be some easy optimizations, I'm not sure.  That function also does way too much, in my opinion, I'd be interested to see if we could replace most usages with much simpler utilities (like `CommCareUser.by_domain`).

esoergel commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -178,6 +178,7 @@ def selected(self):
         return FormManagementMode(self.request.GET.get(self.slug)).mode_name
 
 
+@location_safe
```
:+1: 
Looks like the queries happen in `SubmitHistoryMixin`, which should be location safe already

mkangia replied:
yup! this was just to allow page access to restricted user.

esoergel commented on corehq/apps/data_interfaces/tasks.py:
```diff
@@ -53,6 +54,7 @@ def get_ids_from_url(query_string, domain, couch_user):
         _request.user = couch_user.get_django_user()
         _request.domain = domain
         _request.couch_user.current_domain = domain
+        _request.can_access_all_locations = can_access_all_locations
```
You can also use `couch_user.has_permission(domain, 'access_all_locations')` if you wanted to avoid passing in another parameter.

mkangia replied:
aha! nice. thanks

esoergel commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -498,7 +500,8 @@ def post(self, request, *args, **kwargs):
-            form_ids_or_query_string
+            form_ids_or_query_string,
+            self.request.can_access_all_locations
         )
```
This should probably check permissions on the forms being updated before calling the task.  How does that happen currently?

mkangia replied:
@esoergel i am not sure if something like this already exist and have not added any check as of now.
If need to add one what should be check to ensure all cases are covered. Thinking about:
1. form user_id should be the user's id

esoergel replied:
Yeah, I think the form user ID should come from an accessible user.

mkangia committed:
Avoid inheriting CaseExportScope and rely on CaseListMixin only as the changes in former are now absorbed in latter

mkangia committed:
avoid passing access permission and use couch_user instead

mkangia committed:
avoiding requesting get_all_users_by_domain and fetch all users at once using UserES instead

dimagimon commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -83,12 +88,22 @@ def rows(self):
+                self.request.domain,
+                self.request.couch_user)
+            )
+            accessible_user_ids = user_ids_at_locations(accessible_location_ids)
```
F841 local variable 'accessible_user_ids' is assigned to but never used

mkangia committed:
dimagimon complaint to remove unused variable

dimagimon commented on corehq/apps/data_interfaces/interfaces.py:
```diff
@@ -7,7 +7,10 @@
 
 from corehq.apps.app_manager.const import USERCASE_TYPE
 from corehq.apps.es import cases as case_es
+from corehq.apps.es.users import user_ids_at_locations, UserES
```
F401 'user_ids_at_locations' imported but unused

esoergel reviewed (approved):
cool, looks good to me

esoergel commented:
Oh whoops, except for this issue: https://github.com/dimagi/commcare-hq/pull/14382#discussion_r94460791

mkangia committed:
ensure accessible forms archived

mkangia committed:
Better message on accessing inaccessible forms when managing forms

mkangia commented:
Thank You @esoergel . Added a check at view level to ensure accessible forms accessed. Also moved the logic to fetch form ids from archiving to view. Changes [here](https://github.com/dimagi/commcare-hq/pull/14382/files#diff-9f0c2d0ac3b32ff875ff48f73f7b3e10R500)

esoergel commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -511,6 +519,38 @@ def post(self, request, *args, **kwargs):
             )
         )
 
+    def inaccessible_forms_accessed(self, xform_ids, domain, couch_user):
```
This should have something like this at the top:
```python
if couch_user.has_permission(self.domain, 'access_all_locations'):
    return True
```
so we don't needlessly run this logic (and that query) on non-restricted users.

mkangia replied:
oops! added that earlier but got lost when moving code. thanks ethan.

mkangia committed:
check for inaccessible forms only in case of restricted user when archiving forms

esoergel commented:
:+1:

mkangia committed:
merge mk-243862-locations-submit-history

mkangia committed:
remove duplicate update in context

mkangia committed:
Merge branch 'mk-243862-locations-submit-history' into mk-243862-reassign-cases-manage-forms

mkangia committed:
Add ability for restricted user to assign cases to accessible locations

mkangia commented:
@ethan Post your last review, i just added ability for a restricted user to assign case to accessible locations as well.
Tested all and this looks good. Merging this in base branch.
